torch/nn/parallel/distributed.py

```diff
@@ -1386,8 +1429,18 @@ def _lazy_init(self):
         self._lazy_init_ran = True
 
     def _pre_forward(self, *inputs, **kwargs):
-        if not self._lazy_init_ran:
+        if self._ddp_python_hook and torch._utils.is_compiling():
```

Comment:
wondering- is it cleaner to 
- ditch the `self._ddp_python_hook` flag,
- install the python hook inside _pre_forward under `if is_compiling()`
- not have to worry about removing the python hook under non-compiel mode

Reply:
* We cannot ditch `self._ddp_python_hook` because we could not register the python hook for users who still want to use `DDPOptimizer`. We could not distinguish `torch._dynamo.config.ddp_python_hook` from `torch._dynamo.config.optimize_ddp` by using only `is_compiling()`.

* I tried registering the hook in when both `is_compiing()` and `self._ddp_python_hook` are True. That resulted in recompilation in the second iteration (but none afterward). That's why I change to this design.